wgengine/netstack/netstack.go

```diff
@@ -1041,17 +1042,17 @@ func (ns *Impl) acceptUDP(r *udp.ForwarderRequest) {
 }
 
 func (ns *Impl) handleMagicDNSUDP(srcAddr netip.AddrPort, c *gonet.UDPConn) {
-	// In practice, implementations are advised not to exceed 512 bytes
-	// due to fragmenting. Just to be sure, we bump all the way to the MTU.
-	var maxUDPReqSize = tstun.DefaultMTU()
 	// Packets are being generated by the local host, so there should be
 	// very, very little latency. 150ms was chosen as something of an upper
 	// bound on resource usage, while hopefully still being long enough for
 	// a heavily loaded system.
 	const readDeadline = 150 * time.Millisecond
 
 	defer c.Close()
-	q := make([]byte, maxUDPReqSize)
+	// Implementations are advised not to exceed 512 bytes per DNS
+	// request due to fragmenting but in reality can and do send
+	// much larger packets.
+	q := make([]byte, maxUDPPacketSize)
```

Comment:
So instead of allocating ~1280 bytes here we're allocating ~64K now for DNS queries? The comment says it may exceed 512 but do we really expect DNS queries up to 64KB in a UDP packet?

(Bigger allocs here means more garbage and thus more GCs and thus more CPU usage.)

MagicDNS gets enough traffic to be worth paying attention to performance here probably.

Reply:
Yeah, this is a bummer and I would appreciate suggestions for making it go fast given that in reality, TS_DEBUG_MTU means a UDP packet can be up to ~64K, and when PeerMTU is on, we will regularly see 9000 byte MTU and potentially DNS packets as large as either. Googling a bit suggests that 8K is the real world max of a DNS query but I don't know if that's a hard limit. Thoughts?

Reply:
we could use a sync.Pool of 64KB ones to at least not make garbage in the common/hot cases

Reply:
I was about to suggest `sync.Pool` but also we limit forwarded DNS responses to 4KB
https://github.com/tailscale/tailscale/blob/main/net/dns/resolver/tsdns.go#L42-L45

We can also pick an arbitrary lower limit and respond with the error that tells the client to switch to TCP.

Reply:
e.g.

```patch
diff --git a/wgengine/netstack/netstack.go b/wgengine/netstack/netstack.go
index 54e99180..0217f5e1 100644
--- a/wgengine/netstack/netstack.go
+++ b/wgengine/netstack/netstack.go
@@ -1040,10 +1040,14 @@ func (ns *Impl) acceptUDP(r *udp.ForwarderRequest) {
        go ns.forwardUDP(c, srcAddr, dstAddr)
 }
 
+var udpBufPool = &sync.Pool{
+       New: func() any {
+               b := make([]byte, maxUDPPacketSize)
+               return &b
+       },
+}
+
 func (ns *Impl) handleMagicDNSUDP(srcAddr netip.AddrPort, c *gonet.UDPConn) {
-       // In practice, implementations are advised not to exceed 512 bytes
-       // due to fragmenting. Just to be sure, we bump all the way to the MTU.
-       var maxUDPReqSize = tstun.DefaultMTU()
        // Packets are being generated by the local host, so there should be
        // very, very little latency. 150ms was chosen as something of an upper
        // bound on resource usage, while hopefully still being long enough for
@@ -1051,7 +1055,10 @@ func (ns *Impl) handleMagicDNSUDP(srcAddr netip.AddrPort, c *gonet.UDPConn) {
        const readDeadline = 150 * time.Millisecond
 
        defer c.Close()
-       q := make([]byte, maxUDPReqSize)
+
+       bufp := udpBufPool.Get().(*[]byte)
+       defer udpBufPool.Put(bufp)
+       q := *bufp
 
        // libresolv from glibc is quite adamant that transmitting multiple DNS
        // requests down the same UDP socket is valid. To support this, we read
@@ -1164,7 +1171,11 @@ func startPacketCopy(ctx context.Context, cancel context.CancelFunc, dst net.Pac
        }
        go func() {
                defer cancel() // tear down the other direction's copy
-               pkt := make([]byte, maxUDPPacketSize)
+
+               bufp := udpBufPool.Get().(*[]byte)
+               defer udpBufPool.Put(bufp)
+               pkt := *bufp
+
                for {
                        select {
                        case <-ctx.Done():
```

Reply:
To be honest I don't understand the big picture of how this code is used. Why are we forwarding DNS packets? Who do they come from? Who will be sad if we truncate or drop a packet?

Reply:
Regarding large DNS packets, we have had some problems with responses over 1280 bytes with specific queries and resolvers. About a year ago I diagnosed an issue with pihole and the DNS resolver it was using that was producing abnormally large responses when it had DNSSEC validation enabled. A more recent version stopped doing that, but as an anecdote, this was an in-the-wild case of DNS responses being larger than we could (at the time) handle, and it caused user visible issues (~they couldn't resolve DNSSEC enabled domains with IPv6 addresses).

We have some large buffer pools in wireguard-go already, which we needed to add as part of the packet coalescing work. I wouldn't be too concerned about adding another, though at some point we may need to have bounded sized pools. wireguard-go has an implementation of such a thing here: https://github.com/tailscale/wireguard-go/blob/master/device/pools.go#L13 i don't remember seeing one of our own, though perhaps we have one somewhere.

Reply:
I agree with you both. I added a separate commit co-authored by Brad that adds the buffer pools, then the following commit raises the buffer size.